_posts/2020-05-02-schedule-your-functions.md

```diff
@@ -0,0 +1,264 @@
+
+Now lets go head and create a function which will run on a schedule.
+
+## Lockbot
+
+The `Lockbot` will run once per day at `00:00` and will check for old GitHub issues. The issue will be locked in case it is older than the days we will chose. For our example we can go with 3 months, which is roughly 90 days.
+
+### Get the function template from store
+
+The OpenFaaS framework has rich variety of language templates which you can use to write your own functions. In our example we will use Python 3 with the Debian based template `python3-flask-debian`:
+
+```bash
+faas-cli template store pull python3-flask-debian
+```
+
+### Generate the function
+
+In order to generate the function's backbone we will run the following command where you need to replace `<dockerhub_username>` with your actual username in dockerhub:
+
+```bash
+faas-cli new lockbot --lang python3-flask-debian --prefix <dockerhub_username> && mv lockbot.yml stack.yml
+```
+
+### Generate access token
+
+Go to the [personal access tokens](https://github.com/settings/tokens) page in GitHub and press the `Generate new token` button. The bot will use this to authenticate to the repository you chose.
+
+Now copy the generated token and create `auth-token` secret using the `faas-cli` where you need to replace the `<token>` with the actual copied access token:
+
+```bash
+faas-cli secret create auth-token --from-literal='<token>'
+```
+
+### Configure the function
+
+Open the function's configuration file called `stack.yml`. Append the following lines to the file:
+
+```yml
+    environment:
+      github_username: <github_username>
+      github_repository: <repository_to_watch>
+    secrets:
+      - auth-token
+    annotations:
+      topic: cron-function
+      schedule: "0 0 * * *"
+```
+
+Replace `<repository_to_watch>` with one of your own repositories and `<github_username>` with your GitHub username, which should also own the repository. The whole `stack.yml` file should look something like this:
+
+```yml
+version: 1.0
+provider:
+  name: openfaas
+  gateway: <gateway_url>
+functions:
+  lockbot:
+    lang: python3-flask-debian
+    handler: ./lockbot
+    image: <your_dockerhub_username>/lockbot:latest
+    environment:
+      github_username: <github_username>
+      github_repository: <repository_to_watch>
+      days_before_outdated: 90
+    secrets:
+      - auth-token
+    annotations:
+      topic: cron-function
+      schedule: "0 0 * * *"
+```
+
+### Augment the code
+
+Inside the `lockbot` folder you can see the `handler.py` file which contains the `handle` method which is the entry point to your function. Replace what's inside with the lockbot's code:
+
+```python
+import os
+import requests
+from datetime import datetime
+
+
+def get_difference_in_days(string_date):
```

Comment:
Functions should read like a book, come after they are invoked not before if possible in Python, I believe it is.

Reply:
This is indeed a good rule of thumb which I should keep, but python does not compile so cannot put methods below the calling code

Reply:
You need to add them before calling https://stackoverflow.com/questions/758188/make-function-definition-in-a-python-file-order-independent